File: api/v1/schemas/system_metrics.py
# -*- coding: utf-8 -*-
"""System metrics API schemas."""

from __future__ import annotations

from typing import Any, Dict, List

from pydantic import BaseModel


class SlowOperationItem(BaseModel):
    """One slow operation sample."""

    operation: str
    provider: str
    status: str
    duration_ms: int


class MetricTrendItem(BaseModel):
    """One recent trend bucket for an operation."""

    bucket_ts: int
    count: int
    success: int
    error: int
    cache_hits: int
    cache_misses: int
    total_duration_ms: int
    max_duration_ms: int
    avg_duration_ms: int


class SystemMetricsResponse(BaseModel):
    """Runtime metrics snapshot for API/Web observability."""

    observability: Dict[str, Dict[str, int]]
    observability_by_provider: Dict[str, Dict[str, int]]
    observability_trends: Dict[str, List[MetricTrendItem]]
    recent_slowest: List[SlowOperationItem]
    cache: Dict[str, Dict[str, Any]]


class CacheClearResponse(BaseModel):
    """Response payload for cache clear actions."""

    success: bool
    namespace: str
    cleared_namespaces: List[str]
    cache: Dict[str, Dict[str, Any]]

File: apps/dsa-web/src/api/systemMetrics.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface MetricBucket {
  hits?: number;
  misses?: number;
  writes?: number;
  evictions?: number;
  entries?: number;
  stale?: number;
  clears?: number;
  readErrors?: number;
  count?: number;
  success?: number;
  error?: number;
  cacheHits?: number;
  cacheMisses?: number;
  totalDurationMs?: number;
  maxDurationMs?: number;
  avgDurationMs?: number;
  hitRatePct?: number;
}

export interface SystemMetricsResponse {
  observability: Record<string, MetricBucket>;
  observabilityByProvider: Record<string, MetricBucket>;
  observabilityTrends: Record<string, Array<MetricBucket & { bucketTs: number }>>;
  recentSlowest: Array<{
    operation: string;
    provider: string;
    status: string;
    durationMs: number;
  }>;
  cache: Record<string, Record<string, MetricBucket>>;
}

export interface CacheClearResponse {
  success: boolean;
  namespace: string;
  clearedNamespaces: string[];
  cache: Record<string, Record<string, MetricBucket>>;
}

export const systemMetricsApi = {
  async getMetrics(): Promise<SystemMetricsResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/system/metrics');
    return toCamelCase<SystemMetricsResponse>(response.data);
  },

  async clearCache(namespace: string): Promise<CacheClearResponse> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/system/metrics/cache/clear', null, {
      params: { namespace },
    });
    return toCamelCase<CacheClearResponse>(response.data);
  },
};
